### PR TITLE
Client 6383 client 6372 | Listening to websocket close and checking for signaling state before applying answer

### PR DESCRIPTION
**Pull request by charliesantos**
**Contributing to Twilio**

> All third party contributors acknowledge that any contributions they provide will be made under the same open source license that the open source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### JIRA link(s):

- [CLIENT-6383](https://issues.corp.twilio.com/browse/CLIENT-6383)
- [CLIENT-6372](https://issues.corp.twilio.com/browse/CLIENT-6372)

### Description

- We were using `pstream.on('offline')` to detect network drops which would disconnect the active connection. But signaling server also emits 'offline' event when the token expires, which causes unwanted disconnection. This PR switches to `transport.on('close')` instead which gets emitted on websocket close.
- We are now ignoring iceRestart if signaling state is not `have-local-offer`, which usually happens when a createOffer came in late and a previous answer has already been applied.

### Before review
* [x] Updated CHANGELOG.md if necessary
* [x] Added unit tests if necessary
* [ ] Updated affected documentation
* [ ] Verified locally with `npm run build:release`
* [x] Manually sanity tested running locally
* [x] Ready for review

### Before merge
* [ ] Got one or more +1s
* [ ] Squashed erroneous commits if necessary
* [ ] Re-tested if necessary

### CODE REVIEW

File: lib/twilio/connection.ts [hunk 1]
```diff
@@ -425,7 +425,7 @@ class Connection extends EventEmitter {
 
     // When websocket gets disconnected
     // There's no way to retry this session so we disconnect
-    this.pstream.on('offline', this._disconnect.bind(this));
+    this.pstream.transport.on('close', this._disconnect.bind(this));
```
Comment (ryan-rowland) [inline lib/twilio/connection.ts]: We shouldn’t be reaching into a child here. This means Connection will know about pstream’s implementation details. Maybe we should add a new event to pstream, `close`. WDYT?

File: lib/twilio/rtc/peerconnection.js [hunk 2]
```diff
@@ -698,6 +698,13 @@ PeerConnection.prototype.iceRestart = function() {
         this._removeReconnectionListeners();
         if (!payload.sdp) { return reject(true); }
 
+        // Either this did not come from createOffer from iceRestart
+        // or a createOffer came in late and a previous anwer has already been applied
```
Comment (ryan-rowland) [inline lib/twilio/rtc/peerconnection.js]: sp: answer

File: lib/twilio/connection.ts (hunk 1 above)
Comment (charliesantos) [reply lib/twilio/connection.ts]: Good point, however, pstream already has its own 'close' event which has a different meaning. It's the 'close' event coming from the server, which doesn't necessarily mean websocket's 'close'.

Comment (charliesantos) [reply lib/twilio/connection.ts]: Or, we can still add a close event in pstream and call it something different. Something that's related to transport close, example `transport-close`. WDYT?

Comment (ryan-rowland) [reply lib/twilio/connection.ts]: Man, we should rethink the way we surface these messages. Separate messages received from transport state. Let’s go with `transportClosed` for now.

Comment (charliesantos) [reply lib/twilio/connection.ts]: I agree @ryan-rowland .

**Pull request merged**